src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -1522,40 +1658,31 @@ def _array_notation_rank(array):
 
         :raises NotImplementedError: if the array node does not have any \
                                      children.
-        :raises NotImplementedError: if the Array does not have at least one \
-                                     CodeBlock as a child.
-        :raises NotImplementedError: if each CodeBlock child does not \
-                                     consist entirely of Subscript_Triplets.
-        :raises NotImplementedError: if any of the array slices have bounds \
-                                     (as this is not yet supported).
+
         '''
         if not array.children:
             raise NotImplementedError("An Array reference in the PSyIR must "
                                       "have at least one child but '{0}' has "
                                       "none".format(array.name))
-        # TODO #412. This code will need re-writing once array notation is
-        # supported in the PSyIR (becaues then we won't have any CodeBlocks).
-        cblocks = array.walk(CodeBlock)
-        if not cblocks:
-            raise NotImplementedError(
-                "A PSyIR Array node representing an array access that uses "
-                "Fortran array notation is assumed to have at least one "
-                "CodeBlock as its child but '{0}' has none.".
-                format(array.name))
-        # We only currently support array refs using the colon syntax,
-        # e.g. array(:, :). Each colon is represented in the fparser2
-        # parse tree by a Subscript_Triplet.
+        # Only array refs using basic colon syntax are currently
+        # supported e.g. (a(:,:)).  Each colon is represented in the
+        # PSyIR as a Range node with first argument being an lbound
+        # binary operator, the second argument being a ubound operator
+        # and the third argument being an integer Literal node with
+        # value 1 i.e. a(:,:) is represented as
+        # a(lbound(a,1):ubound(a,1):1,lbound(a,2):ubound(a,2):1) in
+        # the PSyIR.
         num_colons = 0
-        for cblock in cblocks:
-            for stmt in cblock.get_ast_nodes:
-                if not isinstance(stmt, Fortran2003.Subscript_Triplet):
+        for node in array.children:
+            if isinstance(node, Range):
+                # Found array syntax notation. Check that it is the
+                # simple ":" format.
+                try:
+                    _check_range_is_full_extent(node)
+                except NotImplementedError:
                     raise NotImplementedError(
```

Comment:
This is definitely the right place to raise a NotImplementedError. As discussed above, I'm inclined to agree with your misgivings about raising it elsewhere but I think we need to refine the code a bit.

Reply:
I've refined. Hopefully it is to your liking.

Reply:
It is :-)